internal/source/source.go

```diff
@@ -0,0 +1,67 @@
+package source
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"sync"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+)
+
+type Source interface {
+	SourceContent(context.Context, *rukpakv1alpha1.Bundle) (*Result, error)
+}
+
+type Result struct {
+	Bundle         fs.FS
+	ResolvedSource *rukpakv1alpha1.BundleSource
+	State          State
+	Message        string
+}
+
+type State string
+
+const (
+	StatePending   State = "Pending"
+	StateUnpacking State = "Unpacking"
+	StateUnpacked  State = "Unpacked"
+)
+
+type Sourcer struct {
+	sources map[rukpakv1alpha1.BundleSourceType]Source
+	m       sync.RWMutex
+}
+
+var (
+	ErrSourceAlreadyAdded = errors.New("source type already added")
+	ErrSourceNotSupported = errors.New("source type not supported")
+)
+
+func NewSourcer() *Sourcer {
+	return &Sourcer{sources: map[rukpakv1alpha1.BundleSourceType]Source{}}
+}
+
+func (s *Sourcer) Add(typ rukpakv1alpha1.BundleSourceType, source Source) error {
```

Comment:
Making it variadic without a key/value struct would sacrifice compile-time checking since we'd have to make the function `Add(keysAndValues ...interface{})`, which has three runtime ways of going wrong instead of just one:
1. The key already exists (that's a check we would need no matter what)
2. The length of `keysAndValues` is odd
3. The values at even indices are not bundle source types or the values add odd indices are not `Source`s.

There are two other options:
1. Change this to accept a map that we can iterate.
2. Change this to accept a list of `SourcerMapping` objects (or some variation of that name).